src/Executables/CombineH5/CombineH5.cpp

```diff
@@ -48,62 +54,67 @@ void combine_h5(const std::string& file_prefix, const std::string& subfile_name,
-  // std::map
-  for (size_t i = 0; i < file_names.size(); ++i) {
-    h5::H5File<h5::AccessType::ReadWrite> original_file(
-        file_prefix + std::to_string(i) + ".h5", true);
-
-    const auto& original_volume_file =
-        original_file.get<h5::VolumeData>("/" + subfile_name);
-
-    const auto& sorted_volume_data = original_volume_file.get_data_by_element(
-        std::nullopt, std::nullopt, std::nullopt);
-
-    file_map.insert(
-        std::pair<int, std::vector<std::tuple<size_t, double,
-                                              std::vector<ElementVolumeData>>>>(
-            i, sorted_volume_data));
-    original_file.close_current_object();
-  }
-
-  // Append vectors of ElementVolumeData that were split between H5 files for
-  // a given observation id and value
-  for (size_t i = 1; i < file_names.size(); ++i) {
-    for (size_t j = 0; j < file_map.at(0).size(); ++j) {
-      std::get<2>(file_map.at(0)[j])
-          .insert(std::get<2>(file_map.at(0)[j]).end(),
-                  std::get<2>(file_map.at(i)[j]).begin(),
-                  std::get<2>(file_map.at(i)[j]).end());
+  // Obtains list of observation ids to loop over
+  std::vector<size_t> observation_ids =
+      return_observation_ids(file_prefix, subfile_name);
+
+  // Loops over observation ids to write volume data by observation id
+  for (const auto& obs_id : observation_ids) {
+    // Pre-calculates size of vector to store element data and allocates
+    // corresponding memory
+    size_t vector_dim =
+        get_number_of_elements(file_names, subfile_name, obs_id);
+    std::vector<ElementVolumeData> element_data;
+    element_data.reserve(vector_dim);
+
+    double obs_val = 0.0;
+
+    // Loops over input files to append element data into a single vector to be
+    // stored in a single H5
+    for (auto const& file_name : file_names) {
+      h5::H5File<h5::AccessType::ReadOnly> original_file(file_name, false);
+      const auto& original_volume_file =
+          original_file.get<h5::VolumeData>("/" + subfile_name);
+      obs_val = original_volume_file.get_observation_value(obs_id);
+
+      // Get vector of element data for this `obs_id` and `file_name`
+      std::vector<ElementVolumeData> data_by_element =
+          std::move(std::get<2>(original_volume_file.get_data_by_element(
+              obs_val * (1.0 - 4.0 * std::numeric_limits<double>::epsilon()),
+              obs_val * (1.0 + 4.0 * std::numeric_limits<double>::epsilon()),
+              std::nullopt)[0]));
+
+      // Append vector to total vector of element data for this `obs_id`
+      element_data.insert(element_data.end(),
+                          std::make_move_iterator(data_by_element.begin()),
+                          std::make_move_iterator(data_by_element.end()));
+      data_by_element.clear();
+      original_file.close_current_object();
     }
+    auto& new_volume_file = new_file.get<h5::VolumeData>("/" + subfile_name);
```

Comment:
E.g. here you can do a `h5::H5File<h5::AccessType::ReadWrite> new_file(output + "0.h5", true);` The file will get close at the end of the for loop